### PR TITLE
Duplicate argument linter

### LINKED ISSUE #850: A new linter checking duplicate arguments
It is almost always a mistake to call a function with duplicate arguments, i.e. arguments of the same name, like the following:

```r
foo(bar1 = 1, bar1 = 2, bar3 = 3)
```

Therefore, in most cases, it might make sense to check all function calls and ensure there are no duplicate arguments.

One use case of duplicate argument is to produce some table that actually needs duplicate headers for demonstration in documents, e.g.

```r
data <- data.frame(
  mean = 1:5,
  sd = 0.5,
  mean = 2:6,
  sd = 1,
  check.names = FALSE
)
```

### PR DESCRIPTION
**Pull request by renkun-ken**
Closes #850 

This PR implements a `duplicate_argument_linter` to check if there are duplicate arguments supplied to function calls.

### CODE REVIEW

File: R/duplicate_argument_linter.R [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      args_text <- vapply(
+        parse(text = xml2::xml_text(args), keep.source = FALSE),
+        as.character, character(1L)
+      )
+      is_dup_arg <- duplicated(args_text)
+      dup_args <- args[is_dup_arg]
+      if (length(dup_args)) {
+        line1 <- as.integer(xml2::xml_attr(dup_args, "line1"))
+        col1 <- as.integer(xml2::xml_attr(dup_args, "col1"))
+        col2 <- as.integer(xml2::xml_attr(dup_args, "col2"))
```
Comment (AshesITR) [inline R/duplicate_argument_linter.R:35]: Technically,

    tmp <- function(`
    ` = 42) {
    `
    `
    }

is valid R code, so `line2` is not guaranteed to be `line1` for call names.

File: tests/testthat/test-duplicate_argument_linter.R [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    list(message = rex("Duplicate arguments in function call.")),
+    duplicate_argument_linter())
+
+  expect_lint(
+    "list(
+      var = 1,
+      var = 2
+    )",
+    NULL,
+    duplicate_argument_linter(except = "list"))
```
Comment (AshesITR) [inline tests/testthat/test-duplicate_argument_linter.R:64]: Can we test the `except = ...` branch with anonymous functions, too?
This ensures the `calls_names` are detected properly for calls to anonymous functions.

**Review by AshesITR — CHANGES_REQUESTED** [review]
Looks mostly good. Just a few suggestions to improve robustness.
Don't forget a NEWS bullet.

File: R/duplicate_argument_linter.R (hunk 1 above)
Comment (renkun-ken) [reply R/duplicate_argument_linter.R:35]: Here `line1` is only used for reporting. Do you have any suggestion what to do with `line2`? Does lintr support multi-lined lint result?

File: tests/testthat/test-duplicate_argument_linter.R (hunk 2 above)
Comment (renkun-ken) [reply tests/testthat/test-duplicate_argument_linter.R:64]: Yes, exactly.

File: R/duplicate_argument_linter.R (hunk 1 above)
Comment (renkun-ken) [reply R/duplicate_argument_linter.R:35]: The actual range is from `line1, col1` to `line2, col2`.

Comment (AshesITR) [reply R/duplicate_argument_linter.R:35]: IIRC we have some places where `col2` is automatically set to the end of the line if `line2 > line1`. I think that would be fine here, too. It's just a crude edge case anyway as noone in their right mind would use newlines in argument names other than for demonstrational purposes.

File: R/get_source_expressions.R [hunk 3]
```diff
@@ -115,6 +115,28 @@ get_source_expressions <- function(filename, lines = NULL) {
             line = ""
           )
         )
+      } else if (grepl("repeated formal argument", e$message, fixed = TRUE)) {
```
Comment (AshesITR) [inline R/get_source_expressions.R:118]: Very nice :)

File: R/duplicate_argument_linter.R (hunk 1 above)
Comment (renkun-ken) [reply R/duplicate_argument_linter.R]: `col2` tweaked to line ends. Do we need test case for such "crude edge case"?

Comment (AshesITR) [reply R/duplicate_argument_linter.R]: Sounds like a good idea to me.

Comment (renkun-ken) [reply R/duplicate_argument_linter.R]: A crude edge test cases added.

**Review by AshesITR — APPROVED** [review]
LGTM
